thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobResource.java

```diff
@@ -364,4 +382,115 @@ public Response evaluateAlertFilterByFunctionId(@PathParam("functionId") long id
+   * @param tuningJSON the input json string from user
+   * @return full list of all the possible configurations
+   * @throws JSONException
+   */
+  private List<Map<String, String>> listAllTuningParameters(JSONObject tuningJSON) throws JSONException {
+    List<Map<String, String>> tuningParameters = new ArrayList<>();
+    Iterator<String> jsonFieldIterator = tuningJSON.keys();
+    Map<String, List<String>> fieldToParams = new HashMap<>();
+
+    int numPermutations = 1;
+    while(jsonFieldIterator.hasNext()){
+      String field = jsonFieldIterator.next();
+
+      if(field != null && !field.isEmpty()){
+        // JsonArray to String List
+        List<String> params = new ArrayList<>();
+        JSONArray paramArray = tuningJSON.getJSONArray(field);
+        if(paramArray.length() == 0){
+          continue;
+        }
+        for(int i = 0; i < paramArray.length(); i++){
+          params.add(paramArray.get(i).toString());
+        }
+        numPermutations *= params.size();
+        fieldToParams.put(field, params);
+      }
+    }
+
+    if(fieldToParams.size() == 0) { // No possible tuning parameters
+      return tuningParameters;
+    }
+    List<String> fieldList = new ArrayList<>(fieldToParams.keySet());
+    for(int i = 0; i < numPermutations; i++){
+      Map<String, String> combination = new HashMap<>();
+      int index = i;
+      for(String field : fieldList) {
+        List<String> params = fieldToParams.get(field);
+        combination.put(field, params.get(index % params.size()));
+        index /= params.size();
+      }
+      tuningParameters.add(combination);
+    }
+
+    return tuningParameters;
+  }
+
+  /**
+   * Perform anomaly function autotune:
+   *  - run backfill on all possible combinations of tuning parameters
+   *  - keep all the parameter combinations which lie in the goal range
+   *  - return list of parameter combinations along with their performance evaluation
+   * @param functionId
+   * the id of the target anomaly function
+   * @param replayStartTime
+   * the start time of the anomaly function replay in ISO format, e.g. 2017-02-27T00:00:00.000Z
+   * @param replayEndTime
+   * the end time of the anomaly function replay in ISO format
+   * @param timezone
+   * the timezone of replay time
+   * @param tuningJSON
+   * the json object includes all tuning fields and list of parameters
+   * ex: {"baselineLift": [0.9, 0.95, 1, 1.05, 1.1], "baselineSeasonalPeriod": [2, 3, 4]}
+   * @param goal
+   * the expected performance assigned by user
+   * @return
+   * A response containing all satisfied properties with their evaluation result
+   */
+  @POST
+  @Path("function/{id}/autotune")
+  public Response anomalyFunctionAutotune(@PathParam("id") long functionId, @QueryParam("start") String replayStartTime,
+      @QueryParam("end") String replayEndTime, @QueryParam("timezone") String timezone,
+      @QueryParam("tune") String tuningJSON, @QueryParam("goal") double goal,
```

Comment:
can put it as "/autotune/function" to follow similar conventions in alert filter autotune